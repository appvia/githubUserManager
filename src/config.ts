export const config = {
  get addUsers(): boolean {
    return process.env.ADD_USERS?.toLowerCase() === 'true'
  },
  get removeUsers(): boolean {
    return process.env.REMOVE_USERS?.toLowerCase() === 'true'
  },
  get exitCodeOnMissmatch(): number {
    return parseInt(process.env.EXIT_CODE_ON_MISMATCH ?? '0') || 0
  },
  get ignoredUsers(): string[] {
    return process.env.IGNORED_USERS?.toLowerCase().split(',') ?? []
  },
  get githubPrivateKey(): string {
    return Buffer.from(process.env.GITHUB_PRIVATE_KEY, 'base64').toString('utf-8')
  },
  get githubAppID(): number {
    return parseInt(process.env.GITHUB_APP_ID)
  },
  get githubInstallationID(): number {
    return parseInt(process.env.GITHUB_INSTALLATION_ID)
  },
  get githubOrg(): string {
    return process.env.GITHUB_ORG ?? ''
  },
  get googleCredentials(): googleCredentials {
    return JSON.parse(Buffer.from(process.env.GOOGLE_CREDENTIALS, 'base64').toString('utf-8'))
  },
  get googleEmailAddress(): string {
    return process.env.GOOGLE_EMAIL_ADDRESS ?? ''
  },
}

export interface googleCredentials {
  type: string
  project_id: string
  private_key_id: string
  private_key: string
  client_email: string
  client_id: string
  auth_uri: string
  token_uri: string
  auth_provider_x509_cert_url: string
  client_x509_cert_url: string
}
