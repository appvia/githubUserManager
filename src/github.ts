import { createAppAuth } from '@octokit/auth-app'
import { Octokit } from '@octokit/rest'
import * as mod from './github'
import { config } from './config'
import { GetResponseTypeFromEndpointMethod } from '@octokit/types'

export function getAuthenticatedOctokit(): Octokit {
  return new Octokit({
    authStrategy: createAppAuth,
    auth: {
      appId: config.githubAppID,
      privateKey: config.githubPrivateKey,
      installationId: config.githubInstallationID,
    },
  })
}

export async function getGithubUsersFromGithub(): Promise<Set<string>> {
  const octokit = mod.getAuthenticatedOctokit()
  const members = await octokit.paginate(octokit.orgs.listMembers, {
    org: config.githubOrg,
  })

  const pendingInvites = await octokit.paginate(octokit.orgs.listPendingInvitations, {
    org: config.githubOrg,
  })
  const pendingGithubAccounts = formatUserList(pendingInvites)

  const githubAccounts = formatUserList(members)

  if (pendingGithubAccounts.size > 0)
    console.log(`Outstanding GitHub invites for ${Array.from(pendingGithubAccounts).join(', ')}`)

  return new Set([...githubAccounts, ...pendingGithubAccounts])
}

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export function formatUserList(users): Set<string> {
  return new Set(
    users
      .map((user) => user.login?.toLowerCase())
      .flat()
      .filter(Boolean),
  )
}

export async function getUserIdFromUsername(username: string): Promise<number> {
  const octokit = mod.getAuthenticatedOctokit()
  console.log(`Looking up user ${username}`)
  let user
  try {
    user = await octokit.users.getByUsername({ username })
  } catch (error) {
    throw `Unable to find user id for ${username}`
  }
  console.log(`User ${username} userid: ${user.data.id}`)
  return user.data.id
}

export async function addUsersToGitHubOrg(users: Set<string>): Promise<void> {
  for (const user of users) {
    await mod.addUserToGitHubOrg(user)
  }
}

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export async function addUserToGitHubOrg(
  user: string,
): Promise<GetResponseTypeFromEndpointMethod<typeof octokit.orgs.createInvitation> | boolean> {
  const octokit = mod.getAuthenticatedOctokit()
  if (config.ignoredUsers.includes(user.toLowerCase())) {
    console.log(`Ignoring add for ${user}`)
    return false
  }
  const userId = await mod.getUserIdFromUsername(user)
  console.log(`Inviting ${user} (${userId} to ${config.githubOrg})`)
  return await octokit.orgs.createInvitation({
    org: config.githubOrg,
    invitee_id: userId,
  })
}

export async function removeUsersFromGitHubOrg(users: Set<string>): Promise<void> {
  for (const user of users) {
    await mod.removeUserFromGitHubOrg(user)
  }
}

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export async function removeUserFromGitHubOrg(
  user: string,
): Promise<GetResponseTypeFromEndpointMethod<typeof octokit.orgs.removeMembershipForUser> | boolean> {
  const octokit = mod.getAuthenticatedOctokit()
  if (config.ignoredUsers.includes(user.toLowerCase())) {
    console.log(`Ignoring remove for ${user}`)
    return false
  }
  console.log(`Removing user/invitation ${user} from ${config.githubOrg}`)
  return octokit.orgs.removeMembershipForUser({
    org: config.githubOrg,
    username: user,
  })
}
