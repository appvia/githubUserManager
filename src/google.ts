import { google } from 'googleapis'
import * as mod from './google'
import { config } from './config'

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export async function googleAuth() {
  const privatekey = config.googleCredentials
  const jwtClient = new google.auth.JWT(
    privatekey.client_email,
    null,
    privatekey.private_key,
    ['https://www.googleapis.com/auth/admin.directory.user.readonly'],
    config.googleEmailAddress,
  )
  await jwtClient.authorize()
  return jwtClient
}

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export async function getAdminService() {
  return google.admin({
    version: 'directory_v1',
    auth: await googleAuth(),
  })
}

export async function getGithubUsersFromGoogle(): Promise<Set<string>> {
  const service = await mod.getAdminService()

  const userList = await service.users.list({
    customer: 'my_customer',
    maxResults: 250,
    projection: 'custom',
    fields: 'users(customSchemas/Accounts/github(value))',
    customFieldMask: 'Accounts',
  })

  const githubAccounts = mod.formatUserList(userList.data.users)
  return githubAccounts
}

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export function formatUserList(users): Set<string> {
  return new Set(
    users
      .map((user) => user.customSchemas?.Accounts?.github?.map((account) => account.value?.toLowerCase()))
      .flat()
      .filter(Boolean),
  )
}
