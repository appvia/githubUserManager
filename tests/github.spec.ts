jest.mock('@octokit/rest')
import { Octokit } from '@octokit/rest'
import { config } from '../src/config'
import * as mod from '../src/github'

describe('github integration', () => {
  beforeEach(() => {
    jest.spyOn(config, 'githubPrivateKey', 'get').mockReturnValue('helloworld')
    jest.spyOn(config, 'githubAppID', 'get').mockReturnValue(123)
    jest.spyOn(config, 'githubInstallationID', 'get').mockReturnValue(123)
    jest.spyOn(global.console, 'log').mockImplementation()
  })

  it('getAuthenticatedOctokit', () => {
    mod.getAuthenticatedOctokit()
    return expect(Octokit).toMatchSnapshot()
  })
  it('getGithubUsersFromGithub', () => {
    const fakeOctokit = {
      paginate: jest
        .fn()
        .mockResolvedValueOnce([{ login: 'chrisns' }, { login: 'bar' }, { login: 'foo' }])
        .mockResolvedValueOnce([{ login: 'pending' }, { login: 'chrisns' }, { login: 'anotherpending' }]),
      orgs: { listMembers: jest.fn(), listPendingInvitations: jest.fn() },
    }
    // @ts-expect-error mock service isn't a complete implementation, so being lazy and just doing the bare minimum
    jest.spyOn(mod, 'getAuthenticatedOctokit').mockReturnValue(fakeOctokit)
    expect(mod.getGithubUsersFromGithub()).resolves.toMatchSnapshot()
  })
  it('getUserIdFromUsername found', () => {
    const fakeOctokit = {
      users: {
        getByUsername: jest.fn().mockResolvedValue({ data: { id: 123 } }),
      },
    }
    // @ts-expect-error mock service isn't a complete implementation, so being lazy and just doing the bare minimum
    jest.spyOn(mod, 'getAuthenticatedOctokit').mockReturnValue(fakeOctokit)
    return expect(mod.getUserIdFromUsername('foo')).resolves.toMatchSnapshot()
  })

  it('getUserIdFromUsername notfound', () => {
    const fakeOctokit = {
      users: {
        getByUsername: jest.fn().mockRejectedValue(new Error('not found')),
      },
    }
    // @ts-expect-error mock service isn't a complete implementation, so being lazy and just doing the bare minimum
    jest.spyOn(mod, 'getAuthenticatedOctokit').mockReturnValue(fakeOctokit)
    return expect(mod.getUserIdFromUsername('foo')).rejects.toMatchSnapshot()
  })

  it('addUsersToGitHubOrg', async () => {
    const users = new Set(['foo', 'bar'])
    jest.spyOn(mod, 'addUserToGitHubOrg').mockResolvedValue(false)
    await mod.addUsersToGitHubOrg(users)
    return expect(mod.addUserToGitHubOrg).toMatchSnapshot()
  })

  it('removeUsersToGitHubOrg', async () => {
    const users = new Set(['foo', 'bar'])
    jest.spyOn(mod, 'removeUserToGitHubOrg').mockResolvedValue(false)
    await mod.removeUsersToGitHubOrg(users)
    return expect(mod.removeUserToGitHubOrg).toMatchSnapshot()
  })

  it('removeUserToGitHubOrg skip ignore', () => {
    jest.spyOn(config, 'ignoredUsers', 'get').mockReturnValue(['foo'])
    expect(mod.removeUserToGitHubOrg('foo')).resolves.toBe(false)
  })

  it('addUserToGitHubOrg skip ignore', () => {
    jest.spyOn(config, 'ignoredUsers', 'get').mockReturnValue(['foo'])
    expect(mod.addUserToGitHubOrg('foo')).resolves.toBe(false)
  })

  it('addUserToGitHubOrg', async () => {
    const fakeOctokit = {
      orgs: {
        createInvitation: jest.fn().mockResolvedValue(true),
      },
    }
    jest.spyOn(config, 'githubOrg', 'get').mockReturnValue('myorg')
    jest.spyOn(mod, 'getUserIdFromUsername').mockResolvedValue(123)
    // @ts-expect-error mock service isn't a complete implementation, so being lazy and just doing the bare minimum
    jest.spyOn(mod, 'getAuthenticatedOctokit').mockReturnValue(fakeOctokit)
    await mod.addUserToGitHubOrg('foo')
    return expect(fakeOctokit.orgs.createInvitation).toMatchSnapshot()
  })

  it('removeUserToGitHubOrg', async () => {
    const fakeOctokit = {
      orgs: {
        removeMembershipForUser: jest.fn().mockResolvedValue(true),
      },
    }
    jest.spyOn(config, 'githubOrg', 'get').mockReturnValue('myorg')
    jest.spyOn(mod, 'getUserIdFromUsername').mockResolvedValue(123)
    // @ts-expect-error mock service isn't a complete implementation, so being lazy and just doing the bare minimum
    jest.spyOn(mod, 'getAuthenticatedOctokit').mockReturnValue(fakeOctokit)
    await mod.removeUserToGitHubOrg('foo')
    return expect(fakeOctokit.orgs.removeMembershipForUser).toMatchSnapshot()
  })

  it('formatUserList', () => {
    const response = [{ login: 'chrisns' }, { login: 'chrisns' }, { login: 'foo' }, {}]
    return expect(mod.formatUserList(response)).toEqual(new Set(['chrisns', 'foo']))
  })
})
