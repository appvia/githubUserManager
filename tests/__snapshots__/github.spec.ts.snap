// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`github integration addUserToGitHubOrg 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "invitee_id": 123,
        "org": "myorg",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;

exports[`github integration addUsersToGitHubOrg 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "foo",
    ],
    Array [
      "bar",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Promise {},
    },
    Object {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;

exports[`github integration getAuthenticatedOctokit 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "auth": Object {
          "appId": 123,
          "installationId": 123,
          "privateKey": "helloworld",
        },
        "authStrategy": [Function],
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`github integration getGithubUsersFromGithub 1`] = `
Set {
  "chrisns",
  "bar",
  "foo",
  "pending",
  "anotherpending",
}
`;

exports[`github integration getUserIdFromUsername found 1`] = `123`;

exports[`github integration getUserIdFromUsername notfound 1`] = `"Unable to find user id for foo"`;

exports[`github integration removeUserFromGitHubOrg 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "org": "myorg",
        "username": "foo",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;

exports[`github integration removeUsersFromGitHubOrg 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "foo",
    ],
    Array [
      "bar",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Promise {},
    },
    Object {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;
